### DISCUSSION TITLE
Using REST API Dynamic variables to manage offset pagination

### DISCUSSION THREAD

**Discussion by ghost** [Budibase/budibase#8420] [Help] [answered]
Hi all, I'm using an API with offset-based pagination. As Budibase doesn't natively support this type of pagination at the moment, I thought I would use dynamic variable to help navigate between pages. The only issue is I'm not 100% how to access the dynamic variables one they're set, so I'm a bit stuck. 

When the request is made, the results are returned and I store the offset as a dynamic variable (i.e. offset = data.offset). As per the APIs docs, I've included a query parameter and used bindings to link the dynamic variable to this (i.e. apiurl.com?offset={{offset}}) so I can access the next page of results. 

My question is, how do then I trigger the request to reload but this time include the offset value? Right now when I reload the request or refresh the data provider, the first page of results is always returned.

**Comment by melohagan** [comment]
Hey @adamjdavies 

Do you have an app export you could share, or some screenshots to help demonstrate the problem?

**Reply by ghost** [reply]
Hi @melohagan 

Below I've attached a copy of the response I get from the API (it's the Airtable API), where you can clearly see the `offset` variable. 

[Response.txt](https://github.com/Budibase/budibase/files/9954046/Response.txt)

I use a transformer on this data to return it in the format I want:
```
const jsonData = data;
var results = [];
for (var i = 0; i < jsonData.records.length; i++) {
    let counter = jsonData.records[i];
		results.push(counter.fields);
}

return results
```

As you can see in the image, I also use a dynamic variable to store the offset value, which can then be used as a parameter:
<img width="1624" alt="Screenshot 2022-11-07 at 17 27 15" src="https://user-images.githubusercontent.com/64409787/200375548-3dd645b6-9fff-4c81-8b48-3ced4df2bbdf.png">

How would I go about allowing the user to make subsequent API calls that contained the offset value?

Let me know if I can provide further information.

Thanks!

**Comment by melohagan** [comment, marked as answer]
Thanks for the screenshots @adamjdavies 

A Budibase REST connector will always return an array response body because it is designed to be used by tables. Each array element is a row.

I think the dynamic variables work on the transformed API response. So in your case your transformer is not returning the offset.
So you can try changing your JS to be as follows:

```js
const jsonData = data;
var results = [];
for (var i = 0; i < jsonData.records.length; i++) {
    let counter = jsonData.records[i];
    results.push({ ...counter.fields, jsonData.offset });
}

return results
```

And then your dynamic variable would be: `data.0.offset`

**Reply by ghost** [reply]
Hi @melohagan, 

Ah that makes sense. Let me try this and see how it goes. 

Thanks!

**Reply by ghost** [reply]
Hi @melohagan 

Sorry for the delay in getting back to you. I've just tried your suggested edit to my transformer but I'm getting an error: 
`Query Error: Unexpected token '.'`

I'm still fairly new to JS, so I'm not sure what the issue is here. I pasted the snipped into JSFiddle to see where the error may be and this is what it was showing:
<img width="524" alt="image" src="https://user-images.githubusercontent.com/64409787/201966750-adcdd0aa-926e-4d5e-b94d-4bc3f1117f45.png">


 Any ideas?

Thanks!

**Reply by melohagan** [reply]
Ah sorry bad syntax, try this line: `results.push({ ...counter.fields, offset: jsonData.offset });`

**Comment by ghost** [comment]
That's done the job. Thank you!

**Reply by ghost** [reply]
Hey @melohagan 

Me again...

Now that I have the offset value saved as a dynamic value, I'm a bit unsure how best to use it to enable pagination on a table. Having tested a few things, I think it would actually be easier for me to pass the offset value to the api as a regular parameter rather than an dynamic value, and then use a combination of states and conditions to update the data source provider when e.g. a button is clicked. 

Am I overcomplicating things?

Thanks!

**Reply by melohagan** [reply]
There is pagination support in the REST connector, if the API supports pagination: https://docs.budibase.com/docs/rest-pagination

**Reply by ghost** [reply]
Ok, I will try this. Thank you!
